### PR TITLE
Shellcheck shell code

### PR DESCRIPTION
**Pull request by darko-poljak**
Hello!

This is a PR for https://github.com/ungleich/cdist/issues/540.

@4nd3r @telmich @asteven Let us further discuss changes here.
I know there are many changes, but I also turned some of shellcheck notes/warnings/errors off and some are ignored (where it makes sense).

I have also added Makefile target for checking global explorers' and types' shellcode.
`$ make shellcheck`

### CODE REVIEW

File: cdist/conf/type/__staged_file/gencode-local [hunk 1]
```diff
@@ -62,7 +61,7 @@ fetch_file() {
 }
 
 fetch_and_prepare_file() {
-   printf 'tmpdir="$(mktemp -d --tmpdir="/tmp" "%s")"\n' "${__type##*/}.XXXXXXXXXX"
+   printf 'tmpdir="$(TMPDIR="/tmp" mktemp -d "%s")"\n' "${__type##*/}.XXXXXXXXXX"
```
Comment (uqam-fob) [inline cdist/conf/type/__staged_file/gencode-local]: TMPDIR might be set by the user. `TMPDIR=${TMPDIR:-/tmp}` might be more appropriate here.

File: cdist/conf/type/__rvm/explorer/state [hunk 2]
```diff
@@ -28,7 +28,7 @@ if [ "$user" = "root" ]; then
         echo absent
     fi
 else
-    if su - $user -c "[ -d \"\$HOME/.rvm\" ]" ; then
+    if su - "$user" -c "[ -d \"\$HOME/.rvm\" ]" ; then
```
Comment (uqam-fob) [inline cdist/conf/type/__rvm/explorer/state]: `su -` is not supported by all implementations. `su -l` should be used instead

File: cdist/conf/type/__package_pkgng_freebsd/explorer/pkg_version [hunk 3]
```diff
@@ -29,7 +29,7 @@ fi
 
 # Don't produce "no pkgs installed" output -- breaks things
 PKG_OUTPUT=$(pkg info 2>&1)
-echo -n "$(echo "$PKG_OUTPUT" \
+printf "%s" "$(echo "$PKG_OUTPUT" \
```
Comment (uqam-fob) [inline cdist/conf/type/__package_pkgng_freebsd/explorer/pkg_version]: Couldn't this be `printf "%s" "$PKG_OUTPUT" | ...` instead?

File: cdist/conf/type/__package_pkg_freebsd/explorer/pkg_version [hunk 4]
```diff
@@ -30,7 +30,7 @@ fi
 # Don't produce "no pkgs installed" output -- breaks things
 PKG_OUTPUT=$(pkg_info 2>&1)
 if [ ! "$PKG_OUTPUT" = "pkg_info: no packages installed" ]; then
-	echo -n "$(echo "$PKG_OUTPUT" \
+	printf "%s" "$(echo "$PKG_OUTPUT" \
```
Comment (uqam-fob) [inline cdist/conf/type/__package_pkg_freebsd/explorer/pkg_version]: Couldn't this be printf "%s" "$PKG_OUTPUT" | ... instead?

File: cdist/conf/type/__install_reboot/manifest [hunk 5]
```diff
@@ -19,5 +19,6 @@
 #
 
 # set defaults
-options="$(cat "$__object/parameter/options" 2>/dev/null \
-   || echo "" | tee "$__object/parameter/options")"
+# options="$(cat "$__object/parameter/options" 2>/dev/null \
+#   || echo "" | tee "$__object/parameter/options")"
+cat "$__object/parameter/options" 2>/dev/null 1>/dev/null || echo "" > "$__object/parameter/options"
```
Comment (uqam-fob) [inline cdist/conf/type/__install_reboot/manifest]: `[ -f "$__object/parameter/options" ] || touch "$__object/parameter/options"`

File: cdist/conf/type/__install_config/manifest [hunk 6]
```diff
@@ -19,5 +19,7 @@
 #
 
 # set defaults
-chroot="$(cat "$__object/parameter/chroot" 2>/dev/null \
-   || echo "/target" | tee "$__object/parameter/chroot")"
+#chroot="$(cat "$__object/parameter/chroot" 2>/dev/null \
+#   || echo "/target" | tee "$__object/parameter/chroot")"
+
+cat "$__object/parameter/chroot" 2>/dev/null 1>/dev/null || echo "/target" > "$__object/parameter/chroot"
```
Comment (uqam-fob) [inline cdist/conf/type/__install_config/manifest]: [ -f "$__object/parameter/chroot" ] || touch "$__object/parameter/chroot"

File: cdist/conf/type/__docker_compose/gencode-remote [hunk 7]
```diff
@@ -22,9 +22,9 @@
 version="$(cat "$__object/parameter/version")"
 state="$(cat "$__object/parameter/state")"
 
-if [ ${state} = "present" ]; then
+if [ "${state}" = "present" ]; then
     # Download docker-compose file
-        echo 'curl -L "https://github.com/docker/compose/releases/download/'${version}'/docker-compose-$(uname -s)-$(uname -m)" -o /tmp/docker-compose'
+        echo 'curl -L "https://github.com/docker/compose/releases/download/'"${version}"'/docker-compose-$(uname -s)-$(uname -m)" -o /tmp/docker-compose'
```
Comment (uqam-fob) [inline cdist/conf/type/__docker_compose/gencode-remote]: `echo "curl -L 'https://github.com/docker/compose/releases/download/${version}/docker-compose-$(uname -s)-$(uname -m)' -o /tmp/docker-compose"`
would be cleaner imo

File: cdist/conf/explorer/cpu_sockets [hunk 8]
```diff
@@ -25,14 +25,14 @@
 os=$("$__explorer/os")
 case "$os" in
     "macosx")
-        echo "$(system_profiler SPHardwareDataType | grep "Number of Processors" | awk -F': ' '{print $2}')"
+        system_profiler SPHardwareDataType | grep "Number of Processors" | awk -F': ' '{print $2}'
```
Comment (uqam-fob) [inline cdist/conf/explorer/cpu_sockets]: grepping should be done with awk

File: cdist/conf/explorer/cpu_sockets [hunk 9]
```diff
@@ -25,14 +25,14 @@
...
     ;;
 
     *)
     if [ -r /proc/cpuinfo ]; then
         sockets="$(grep "physical id" /proc/cpuinfo | sort | uniq | wc -l)"
-        if [ ${sockets} -eq 0 ]; then
-            sockets="$(cat /proc/cpuinfo | grep "processor" | wc -l)"
+        if [ "${sockets}" -eq 0 ]; then
+            sockets="$(grep -c "processor" /proc/cpuinfo)"
```
Comment (uqam-fob) [inline cdist/conf/explorer/cpu_sockets]: quotes are not needed around subshell calls

File: cdist/conf/type/__ccollect_source/manifest [hunk 10]
```diff
@@ -22,7 +22,7 @@ name="$__object_id"
 state="$(cat "$__object/parameter/state")"
 source="$(cat "$__object/parameter/source")"
 destination="$(cat "$__object/parameter/destination")"
-ccollectconf="$(cat "$__object/parameter/ccollectconf" | sed 's,/$,,')"
+ccollectconf="$(sed 's,/$,,' "$__object/parameter/ccollectconf")"
```
Comment (uqam-fob) [inline cdist/conf/type/__ccollect_source/manifest]: `sed -E 's,/+$,,'` is what is wanted

**Comment by uqam-fob** [comment]
Quoting is really messed up. many constructs like `echo "\"$x\""`. Also variables are not always quoted; notably in for loops (`for e in $__object/*; do`).

**Comment by darko-poljak** [comment]
@uqam-fob Thanks.
Stay tuned for the next iteration.

**Comment by darko-poljak** [comment]
@uqam-fob I have applied your suggestions. If you have time to review more, I will be very thankful. :)

File: cdist/conf/explorer/cpu_sockets [hunk 11]
```diff
@@ -25,14 +25,14 @@
...
 case "$os" in
     "macosx")
-        system_profiler SPHardwareDataType | grep "Number of Processors" | awk -F': ' '{print $2}'
+        system_profiler SPHardwareDataType | grep "Number of Processors" | awk -F ': ' '/Number of Processors/ {print $2}'
     ;;
 
     *)
     if [ -r /proc/cpuinfo ]; then
-        sockets="$(grep "physical id" /proc/cpuinfo | sort | uniq | wc -l)"
+        sockets=$(grep "physical id" /proc/cpuinfo | sort -u | wc -l)
```
Comment (uqam-fob) [inline cdist/conf/explorer/cpu_sockets]: `sort -u` is not defined by POSIX
http://pubs.opengroup.org/onlinepubs/9699919799/utilities/sort.html

**Comment by uqam-fob** [comment]
Reviewed the last 2 commits, looks good.

I suspect the issues pointed out in https://github.com/ungleich/cdist/pull/584#issuecomment-345099992 were willfully ignored. Here are a few examples; to make sure we are on the same page.

messy quotes:
- https://github.com/ungleich/cdist/pull/584/files#diff-017a05277591ae0ffe013e27ce23b4baR35
- https://github.com/ungleich/cdist/pull/584/files#diff-a144abe3e51522906ebab325e5abb39dR45
- https://github.com/ungleich/cdist/pull/584/files#diff-a144abe3e51522906ebab325e5abb39dR83

non-quoted variables:
- https://github.com/ungleich/cdist/pull/584/files#diff-7f11a4d51eff969759aab4771e6a50aaR22
- https://github.com/ungleich/cdist/pull/584/files#diff-88ee4888fe6c430267dc384f1a576045R87
- https://github.com/ungleich/cdist/pull/584/files#diff-7ad707301feb599d2553978ec4719047R53

Also, please note that I stopped commenting on repeated issues. For the sake of clarity, here a other occurences of points I previously mentionned (again, non-exhaustive):
- https://github.com/ungleich/cdist/pull/584/files#diff-a144abe3e51522906ebab325e5abb39dR69
- https://github.com/ungleich/cdist/pull/584/files#diff-9388eac7ad889cea12502f30dae10f29R70
- https://github.com/ungleich/cdist/pull/584/files#diff-0561c4dbf5d3d67aa8dcc62b0d239bc4R3
- https://github.com/ungleich/cdist/pull/584/files#diff-321b2665ee268dacef839c81011f09beR58

But, again, nice work.

File: cdist/conf/explorer/cpu_sockets (hunk 11 above)
Comment (darko-poljak) [reply cdist/conf/explorer/cpu_sockets]: @uqam-fob Hm... But there it is -u option. Am I getting something wrong?

**Comment by darko-poljak** [comment]
@uqam-fob Well, it was late and could not go through all of them :) But I will at some time soon.

**Comment by darko-poljak** [comment]
@uqam-fob Here are some additional commits. Some quoting I don't want to refactor, for now, because some code is complex. And that would mean that we have to test every type again.
I am not sure if, with current state of this branch, nothing is broken.

File: cdist/conf/explorer/cpu_sockets (hunk 11 above)
Comment (uqam-fob) [reply cdist/conf/explorer/cpu_sockets]: Sorry for the confusion! You're absolutely right. I did not notice that `u` in the corner.

File: cdist/conf/type/__cron/gencode-remote [hunk 12]
```diff
@@ -79,18 +79,24 @@ DONE
 case "$state_should" in
     present)
         # if we insert new entry, filter also all entrys out with the same id 
-        echo "("
-        echo "crontab -u $user -l 2>/dev/null | grep -v -E \"$filter\" | grep -v \"# $name\\$\" 2>/dev/null || true"
-        echo "echo '$entry'"
-        echo ") | crontab -u $user -"
+        cat <<EOF
+(
+crontab -u $user -l 2>/dev/null | grep -v -E '$filter' | grep -v '# $name\\$' 2>/dev/null || true
```
Comment (uqam-fob) [inline cdist/conf/type/__cron/gencode-remote]: `$user` should probably be single-quoted

File: cdist/conf/type/__firewalld_start/gencode-remote [hunk 13]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh -e
+"#!/bin/sh -2
```
Comment (uqam-fob) [inline cdist/conf/type/__firewalld_start/gencode-remote]: There seems to be something I am missing here. Why does the line begin with a double-quote and what does `-2` means%

File: cdist/conf/type/__jail_freebsd10/gencode-remote [hunk 14]
```diff
@@ -130,11 +130,11 @@ deleteJail() {
...
+   echo "rm -rf '${jaildir}/rw/${name}'"
 # Remove the jail directory
-   echo "rm -rf \"${jaildir}/${name}\""
+   echo "rm -rf '${jaildir}/${name}'"
 # Remove the jail's fstab
-   echo "rm -f \"/etc/fstab.${name}\""
+   echo "rm -f '/etc/fstab.${name}'"
 # Remove jail entry from jail.conf
    cat <<EOF
       sed -i .bak -E -e "/^${name} {\$/,/^}\\\$/d" /etc/jail.conf
```
Comment (uqam-fob) [inline cdist/conf/type/__jail_freebsd10/gencode-remote]: `sed -i` is not defined by POSIX (double-checked ;) http://pubs.opengroup.org/onlinepubs/9699919799/utilities/sed.html

The POSIX-compliant way to do inlines is to copy the file and sed it back to its original location.

File: cdist/conf/type/__package_pip/gencode-remote [hunk 15]
```diff
@@ -53,15 +53,15 @@ case "$state_should" in
     present)
         if [ "$runas" ]
         then
-            echo "su -c \"$pip install -q $name\" $runas"
+            echo "su -c '$pip install -q $name' $runas"
```
Comment (uqam-fob) [inline cdist/conf/type/__package_pip/gencode-remote]: `$runas` should probably be quoted

File: cdist/conf/type/__postgres_database/explorer/state [hunk 16]
```diff
@@ -20,7 +20,7 @@
 
 name="$__object_id"
 
-if su - postgres -c "echo '\q' | psql '$name'" 2>/dev/null; then
+if su -l postgres -c "echo '\q' | psql '$name'" 2>/dev/null; then
```
Comment (uqam-fob) [inline cdist/conf/type/__postgres_database/explorer/state]: `psql` accepts commands with `-c`
Additionally, command options should always be before other arguments. Only GNU's `getopt(3)` transparently parses all arguments for options (until `--`). BSDs, among others, stop searching for options at the first non-option argument.

`su -l -c "psql -c '\\q' '$name'" 2>/dev/null` should do the trick

File: cdist/conf/type/__postgres_database/gencode-remote [hunk 17]
```diff
@@ -29,10 +29,10 @@ if [ "$state_should" != "$state_is" ]; then
          if [ -f "$__object/parameter/owner" ]; then
             owner="-O '$(cat "$__object/parameter/owner")'"
          fi
-         echo "su - postgres -c \"createdb $owner '$name'\""
+         echo "su -l postgres -c \"createdb $owner '$name'\""
```
Comment (uqam-fob) [inline cdist/conf/type/__postgres_database/gencode-remote]: See https://github.com/ungleich/cdist/pull/584/commits/1143b448b1c241add08592bf249e69d786aac45c#r152696289

File: cdist/conf/type/__rvm/gencode-remote [hunk 18]
```diff
@@ -28,12 +28,12 @@ state_should="$(cat "$__object/parameter/state")"
 case "$state_should" in
     present)
         cat << DONE
-su - $user -c "unset rvm_path; unset rvm_bin_path; unset rvm_prefix; unset rvm_version; curl -L get.rvm.io | bash -s stable"
+su -l $user -c "unset rvm_path; unset rvm_bin_path; unset rvm_prefix; unset rvm_version; curl -L get.rvm.io | bash -s stable"
```
Comment (uqam-fob) [inline cdist/conf/type/__rvm/gencode-remote]: See https://github.com/ungleich/cdist/pull/584/commits/1143b448b1c241add08592bf249e69d786aac45c#r152696289

File: cdist/conf/type/__firewalld_start/gencode-remote (hunk 13 above)
Comment (darko-poljak) [reply cdist/conf/type/__firewalld_start/gencode-remote]: @uqam-fob It is a vim typo. I fixed it.

File: cdist/conf/type/__postgres_database/explorer/state (hunk 16 above)
Comment (darko-poljak) [reply cdist/conf/type/__postgres_database/explorer/state]: For example, FreeBSD su does not have -c option for running command.
-c defines settings of the specified login class.
su synopsis: su [-] [-c class] [-flms] [login [args]]
args are passed to the login shell of the target login.
I took a look at OpenBSD and NetBSD su man pages and they work in that way too.

I think this form will work for GNU and BSDs, at least.
Do you have some other suggestion?

File: cdist/conf/type/__apt_ppa/manifest [hunk 19]
```diff
@@ -18,7 +18,7 @@
 # along with cdist. If not, see <http://www.gnu.org/licenses/>.
 #
 
-name="$__object_id"
+# name="$__object_id"
```
Comment (tom-ee) [inline cdist/conf/type/__apt_ppa/manifest]: any reason to keep this as a comment?

File: cdist/conf/type/__directory/gencode-remote [hunk 20]
```diff
@@ -105,7 +105,7 @@ case "$state_should" in
    ;;
    absent)
         if [ "$type" = "directory" ]; then
-            echo rm -rf \"$destination\"
+            echo "rm -rf '$destination'"
             echo remove >> "$__messages_out"
```
Comment (tom-ee) [inline cdist/conf/type/__directory/gencode-remote]: double-quote "remove" to be consistent?

File: cdist/conf/type/__install_mkfs/manifest [hunk 21]
```diff
@@ -19,13 +19,18 @@
...
+# else
+#    device="/$__object_id"
+#    echo "$device" > "$__object/parameter/device"
+# fi
+
+# type="(cat "$__object/parameter/type")"
+
+# options="(cat "$__object/parameter/options")"
+
+if [ ! -f "$__object/parameter/device" ]; then
```
Comment (tom-ee) [inline cdist/conf/type/__install_mkfs/manifest]: Does this only implement the `else`-case of the previous version?

File: cdist/conf/type/__install_partition_msdos_apply/gencode-remote [hunk 22]
```diff
@@ -21,7 +21,7 @@
 #set -x
 
 die() {
-   echo "[__install_partition_msdos_apply] $@" >&2
+   echo "[__install_partition_msdos_apply]" "$@" >&2
```
Comment (tom-ee) [inline cdist/conf/type/__install_partition_msdos_apply/gencode-remote]: New version does not have the space-character (" ") after the `]`

File: cdist/conf/type/__install_mkfs/manifest (hunk 21 above)
Comment (tom-ee) [reply cdist/conf/type/__install_mkfs/manifest]: Ah, never mind. This just _sets_ the options for gencode-remote.

File: cdist/conf/type/__jail_freebsd10/gencode-remote [hunk 23]
```diff
@@ -36,7 +36,7 @@ state="$(cat "$__object/parameter/state")"
 
 started="true"
 # If the user wants the jail gone, it implies it shouldn't be started.
-[ -f "$__object/parameter/stopped" -o "$state" = "absent" ] && started="false"
+([ -f "$__object/parameter/stopped" ] || [ "$state" = "absent" ]) && started="false"
```
Comment (tom-ee) [inline cdist/conf/type/__jail_freebsd10/gencode-remote]: we can avoid the sub-shell by using command grouping `{ true; false; }`:

    { [ -f "$__object/parameter/stopped" ] || [ "$state" = "absent" ]; } && started="false"

 (there are other occurrences of this)

**Comment by darko-poljak** [comment]
This PR is in some kind of stale state.
I will close it and create new ungleich/shellcheck branch.
It also has too many changes.
I suggest to start doing it step by step, in little changes.
Too many changes at once could break too many things.
We expect PRs on this subject.

**Pull request closed without merging**